confluent_kafka/schema_registry/__init__.py

```diff
@@ -20,3 +20,139 @@
 # from confluent_kafka.serialization.serdes import RecordNameStrategy, \
 #     TopicNameStrategy, TopicRecordNameStrategy
 #
+import io
+
+_MAGIC_BYTE = 0
+
+
+class MessageField(object):
+    """
+    Enum like object for identifying Message fields.
+
+    Attributes:
+        NONE (int): None
+        KEY (int): Message key
+        VALUE (int): Message value
+
+    """
+    NONE = 0
+    KEY = 1
+    VALUE = 2
+    _str = ("none", "key", "value")
+
+    @staticmethod
+    def __str__(field):
+        """
+        Returns a string representation for a MessageField value.
+
+        Args:
+            field (MessageField): MessageField value
+
+        Returns:
+             String value for MessageField
+
+        """
+        return MessageField._str[field]
+
+
+class SerializationContext(object):
+    """
+    SerializationContext provides additional context to the serializer about
+    the data it's serializing.
+
+    Args:
+        - topic (str): Topic the serialized data will be sent to
+        - field (MessageField): Describes what part of the message is
+            being serialized.
+
+    """
+    __slots__ = ["topic", "field"]
+
+    def __init__(self, topic, field):
+        self.topic = topic
+        self.field = field
+
+
+class SubjectNameStrategy(object):
+    """
+    Instructs the Serializer on how to construct subject names when registering
+    schemas. See the Schema Registry documentation for additional details.
+
+    .. _SchemaRegistry Documentation:
+        https://docs.confluent.io/current/schema-registry/serializer-formatter.html#group-by-topic-or-other-relationships
+
+    """
+    def __call__(self, schema, ctx):
+        raise NotImplementedError
+
+    @classmethod
+    def __str__(cls, strategy):
+        return cls._str[strategy]
+
+
+class TopicNameStrategy(SubjectNameStrategy):
+    """
+    Constructs a subject name in the form of {topic}={key|value}.
+
+    Args:
+        schema (Schema): Schema associated with a subject
+        ctx (SerializationContext): Serialization context.
+
+    """
+
+    def __call__(self, schema, ctx):
+        return ctx.topic + "-" + str(ctx.field)
+
+
+class TopicRecordNameStrategy(SubjectNameStrategy):
+    """
+    Constructs a subject name in the form of {topic}={record.name}.
+
+    Args:
+        schema (Schema): Schema associated with a subject
+        ctx (SerializationContext): Serialization context.
+
+    """
+
+    def __call__(self, schema, ctx):
+        return ctx.topic + "-" + schema.name
+
+
+class RecordNameStrategy(SubjectNameStrategy):
+    """
+    Constructs a subject name in the form of {record.name}.
+
+    Args:
+        schema (Schema): Schema associated with a subject
+        ctx (SerializationContext): Serialization context.
+
+    """
+
+    def __call__(self, schema, ctx):
+        return schema.name
+
+
+class ReferenceSubjectNameStrategy(SubjectNameStrategy):
```

Comment:
ReferenceSubjectNameStrategy is not a SubjectNameStrategy.

most correctly, you should have a ReferenceNameStrategy class derive from a base ReferenceSubjectNameStrategy.